
# WSI laboratorium 4 
##### -Mateusz Matukiewicz 325195
## Maszyny wektorów nośnych (SVM)

Wykorzystywane metody jądrowe:
- linowa
- rbf

Dostosowanie zbioru danych do problemu klasyfikacji binarnej:
* Wino jest uznawane za "dobre", gdy jego jakość 
jest conajmniej równa 6, inaczej wino uznajemy za "złe"
* Zbiór danych został podzielony na zbiór trenujący oraz testowy 
80% zbiór trenujący oraz 20% zbiór testujący

## Badanie wpływu hiperparametrów na algorytm SVM

### Hiperparemetry:
- C (Parametr reguralyzacji)
- Kernel Function (Metoda jądrowa -> Liniowa lub RBF)
- Sigma (Przy metodzie jądrowej RBF)
- Learning Rate (Rozmiar kroku używany w gradiencie)
- Iterations (Iteracje przez zbior danych podczas trenowania)

Wyniki Przedstawione będą w tabelkach.

In [2]:
from enum import Enum
from svm import SVM, optim_params, PrepareData, kernel_types
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
import pandas as pd

In [3]:
data = PrepareData()
X_train, X_test, y_train, y_test = train_test_split(np.array(data.X),np.array(data.y), test_size=0.2, random_state=18)
y_train, y_test = data.wrap_targets(y_train,y_test)


# Funkcja jądrowa - liniowa

### Zmienna C

In [4]:
accuracy = []
c_params = [0.1, 2, 10]
for c in c_params:
    op_linear = optim_params(C=c, kernel=kernel_types.LINEAR,lr=1e-3, iters=100)
    svm = SVM(op_linear)
    svm.fit(X_train, y_train)

    predicts = svm.predict(X_test)
    accuracy.append(accuracy_score(y_test, predicts))


result = pd.DataFrame(
    {
        'C': c_params,
        'learning_rate': [1e-3 for _ in range(3)],
        'iterations': [100 for _ in range(3)],
        'accuracy': accuracy })

result

C  learning_rate  iterations  accuracy
0   0.1          0.001         100      0.64
1   2.0          0.001         100      0.64
2  10.0          0.001         100      0.64

### Wspołczynnik uczenia 
(Learning rate)

In [5]:
accuracy = []
lr_params = [1e-10, 1e-4, 1]
for lrate in lr_params:
    op_linear = optim_params(C=1.0, kernel=kernel_types.LINEAR,lr=lrate, iters=100)
    svm = SVM(op_linear)
    svm.fit(X_train, y_train)

    predicts = svm.predict(X_test)
    accuracy.append(accuracy_score(y_test, predicts))


result = pd.DataFrame(
    {
        'C': [1.0 for _ in range(3)],
        'learning_rate': lr_params,
        'iterations': [100 for _ in range(3)],
        'accuracy': accuracy })

result

C  learning_rate  iterations  accuracy
0  1.0   1.000000e-10         100  0.506154
1  1.0   1.000000e-04         100  0.640000
2  1.0   1.000000e+00         100  0.640000

### Iteracje
Iterations

In [6]:
accuracy = []
iters_params = [100, 300, 500]
for iter in iters_params:
    op_linear = optim_params(C=1.0, kernel=kernel_types.LINEAR,lr=1e-3, iters=iter)
    svm = SVM(op_linear)
    svm.fit(X_train, y_train)

    predicts = svm.predict(X_test)
    accuracy.append(accuracy_score(y_test, predicts))


result = pd.DataFrame(
    {
        'C': [1.0 for _ in range(3)],
        'learning_rate': [1e-3 for _ in range(3)],
        'iterations': iters_params,
        'accuracy': accuracy })

result

C  learning_rate  iterations  accuracy
0  1.0          0.001         100      0.64
1  1.0          0.001         300      0.64
2  1.0          0.001         500      0.64

# Funkcja jądrowa RBF

### Zmienna C

In [7]:
accuracy = []
c_params = [0.1, 2, 10]
for c in c_params:
    op_rbf = optim_params(C=c, kernel=kernel_types.RBF,lr=1e-3, iters=100, sigma=0.5)
    svm = SVM(op_rbf)
    svm.fit(X_train, y_train)

    predicts = svm.predict(X_test)
    accuracy.append(accuracy_score(y_test, predicts))


result = pd.DataFrame(
    {
        'C': c_params,
        'sigma': [0.5 for _ in range(3)],
        'learning_rate': [1e-3 for _ in range(3)],
        'iterations': [100 for _ in range(3)],
        'accuracy': accuracy })

result

C  sigma  learning_rate  iterations  accuracy
0   0.1    0.5          0.001         100  0.656154
1   2.0    0.5          0.001         100  0.735385
2  10.0    0.5          0.001         100  0.734615

### Zmienna sigma
W funkcji rbf

In [8]:
accuracy = []
sigma_params =  [0.1  , 0.5 , 2  ]
for sig in sigma_params:
    op_rbf = optim_params(C=1.0, kernel=kernel_types.RBF,lr=1e-4, iters=100, sigma=sig)
    svm = SVM(op_rbf)
    svm.fit(X_train, y_train)

    predicts = svm.predict(X_test)
    accuracy.append(accuracy_score(y_test, predicts))


result = pd.DataFrame(
    {
        'C': [1.0 for _ in range(3)],
        'sigma': sigma_params,
        'learning_rate': [1e-4 for _ in range(3)],
        'iterations': [100 for _ in range(3)],
        'accuracy': accuracy })


result

C  sigma  learning_rate  iterations  accuracy
0  1.0    0.1         0.0001         100  0.733077
1  1.0    0.5         0.0001         100  0.731538
2  1.0    2.0         0.0001         100  0.660000

### Współczynik uczenia
Learning rate

In [9]:
accuracy = []
lr_params = [1e-10, 1e-4, 1]
for lrate in lr_params:
    op_rbf = optim_params(C=1.0, kernel=kernel_types.RBF,lr=lrate, iters=100, sigma=0.5)
    svm = SVM(op_rbf)
    svm.fit(X_train, y_train)

    predicts = svm.predict(X_test)
    accuracy.append(accuracy_score(y_test, predicts))


result = pd.DataFrame(
    {
        'C': [1.0 for _ in range(3)],
        'sigma': [0.5 for _ in range(3)],
        'learning_rate': lr_params,
        'iterations': [100 for _ in range(3)],
        'accuracy': accuracy })

result

C  sigma  learning_rate  iterations  accuracy
0  1.0    0.5   1.000000e-10         100  0.734615
1  1.0    0.5   1.000000e-04         100  0.733846
2  1.0    0.5   1.000000e+00         100  0.561538

### Iteracje
Iterations

In [10]:
accuracy = []
iters_params = [100  , 300 , 500]
for iter in iters_params:
    op_rbf = optim_params(C=1.0, kernel=kernel_types.RBF,lr=1e-3, iters=iter, sigma=0.5)
    svm = SVM(op_rbf)
    svm.fit(X_train, y_train)

    predicts = svm.predict(X_test)
    accuracy.append(accuracy_score(y_test, predicts))

result = pd.DataFrame(
    {
        'C': [1.0 for _ in range(3)],
        'sigma': [0.5 for _ in range(3)],
        'learning_rate': [1e-3 for _ in range(3)],
        'iterations': iters_params,
        'accuracy': accuracy})

result

C  sigma  learning_rate  iterations  accuracy
0  1.0    0.5          0.001         100  0.734615
1  1.0    0.5          0.001         300  0.736154
2  1.0    0.5          0.001         500  0.733846

## Wnioski

Dokładność dla zbadanych parametrów w przypadku funkcji jądrowej liniowej oscyluje w okolicach 64%, przy czym najgorszy wynik uzyskałem dla najmniejszego współczynnika uczenia = 1e-10, natomiast w przypadku funckji jądrowej rbf dokladność oscyluje w okolicach 73%, przy czym najgorszy wynik uzyskałem przy testowaniu współczynnika uczenia = 1 (56%).

Wnioski ogólne:

Wybór odpowiednich parametrów, takich jak współczynniki uczenia, jest kluczowy dla osiągnięcia optymalnej dokładności modelu.

Dodatkowo możemy zauważyć, że funckja jądrowa rbf jest bardziej opytmalna do zadanego dataset-u niż funkcja liniowa, lecz wymaga ona starannej regulacji parametrów.